src/app/backend-api.service.ts

```diff
@@ -949,54 +951,129 @@ export class BackendApiService {
+            const localNode = this.GetStorage(this.LastLocalNodeKey);
+            return this.RegisterGroupMessagingKey(
+              localNode,
+              SenderPublicKeyBase58Check,
+              res.messagingPublicKeyBase58Check,
+              'default-key',
+              res.messagingKeySignature,
+              [],
+              {},
+              MinFeeRateNanosPerKB
+            );
+          };
+
+          const localNode = this.GetStorage(this.LastLocalNodeKey);
+
+          const launchDefaultMessagingKey$ = () =>
+            this.identityService
+              .launchDefaultMessagingKey(SenderPublicKeyBase58Check)
+              .pipe(timeout(30000));
+
+          const submitEncryptedMessage$ = (encrypted: any) => {
+            // Now we will use the ciphertext encrypted to user's messaging keys as part of the metadata of the
+            // sendMessage transaction.
+            const EncryptedMessageText = encrypted.encryptedMessage;
+            // Determine whether to use V3 messaging group key names for sender or recipient.
+            const senderV3 = partyMessagingKeys.IsSenderMessagingKey;
+            const SenderMessagingGroupKeyName = senderV3
+              ? partyMessagingKeys.SenderMessagingKeyName
+              : '';
+            const recipientV3 = partyMessagingKeys.IsRecipientMessagingKey;
+            const RecipientMessagingGroupKeyName = recipientV3
+              ? partyMessagingKeys.RecipientMessagingKeyName
+              : '';
+            return this.post(
+              endpoint,
+              BackendRoutes.RoutePathSendMessageStateless,
+              {
+                SenderPublicKeyBase58Check,
+                RecipientPublicKeyBase58Check,
+                EncryptedMessageText,
+                SenderMessagingGroupKeyName,
+                RecipientMessagingGroupKeyName,
+                MinFeeRateNanosPerKB,
+              }
+            ).pipe(
+              map((request) => {
+                return { ...request };
+              })
+            );
+          };
+          // call encrypt and see what happens
+          return callEncrypt$().pipe(
+            switchMap((res: any) => {
+              // Verify we have the messaging key
+              return of({
+                isMissingRandomness:
+                  res?.encryptedMessage === '' &&
+                  res?.requiresEncryptedMessagingKeyRandomness,
+                res,
+              });
+            }),
+            switchMap(({ isMissingRandomness, res }) => {
+              if (!isMissingRandomness) {
+                console.log('success case');
+                // easy pz return early
+                return submitEncryptedMessage$(res);
+              } else {
+                // otherwise, launch
+                return launchDefaultMessagingKey$().pipe(
+                  switchMap((res) => {
+                    console.log(res);
+                    if (res.encryptedToApplicationGroupMessagingPrivateKey) {
```

Comment:
so unfortunately I think we'll need to update identity to return the `encryptedMessaingKeyRandomness` in the response from messaging window API